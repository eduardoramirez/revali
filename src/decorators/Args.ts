import {registrar} from 'revali/metadata'
import {EmptyConstructor} from 'revali/types'

export function Args() {
  return (target: EmptyConstructor<any>) => {
    registrar.storeIsArgs(target)
  }
}
