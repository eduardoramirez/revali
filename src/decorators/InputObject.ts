import {registrar} from 'revali/metadata'
import {EmptyConstructor} from 'revali/types'

export interface InputObjectTypeConfig {
  name?: string
  description?: string
}

export function InputObjectType(config: InputObjectTypeConfig = {}) {
  return (target: EmptyConstructor<any>) => {
    registrar.storeInputObjectMetadata(target, {...config, name: config.name || target.name})
  }
}
