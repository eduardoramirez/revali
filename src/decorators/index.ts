export * from 'revali/decorators/Arg'
export * from 'revali/decorators/Args'
export * from 'revali/decorators/Field'
export * from 'revali/decorators/Implements'
export * from 'revali/decorators/InputField'
export * from 'revali/decorators/InputObject'
export * from 'revali/decorators/Interface'
export * from 'revali/decorators/Object'
