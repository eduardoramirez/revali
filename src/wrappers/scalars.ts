import {GraphQLBoolean, GraphQLFloat, GraphQLID, GraphQLInt, GraphQLString} from 'graphql'

import {wrapScalar} from 'revali/wrappers/Wrapper'

export const TSGraphQLBoolean = wrapScalar<boolean>(GraphQLBoolean)
export const TSGraphQLFloat = wrapScalar<number>(GraphQLFloat)
export const TSGraphQLID = wrapScalar<number | string>(GraphQLID)
export const TSGraphQLInt = wrapScalar<number>(GraphQLInt)
export const TSGraphQLString = wrapScalar<string>(GraphQLString)
