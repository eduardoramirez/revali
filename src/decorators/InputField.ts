import {GraphQLInputType, Thunk} from 'graphql'

import {graph} from 'revali/graph'
import {Constructor, Omit} from 'revali/types'
import {resolveThunk, resolveType} from 'revali/utils'
import {WrapperOrType} from 'revali/wrappers/Wrapper'

export interface InputFieldOptions<TValue> {
  type: WrapperOrType<TValue, GraphQLInputType>
  defaultValue?: TValue
  description?: string
}

export type PrimitiveInputFieldConfig<TValue> = Omit<InputFieldOptions<TValue>, 'type'>

const getDefaultValueFromPrototype = (prototype: Record<any, any>, key: string) => {
  const Args = prototype.constructor as Constructor<any>
  if (!Args || typeof Args !== 'function' || Args.length > 0) {
    return undefined
  }
  return new Args()[key]
}

function InputFieldDecorator<TValue>(config?: Thunk<Partial<InputFieldOptions<TValue>>>) {
  return <TName extends string>(prototype: Record<TName, TValue>, key: TName) => {
    graph.createInputField(prototype.constructor, () => {
      const resolved = (config && resolveThunk(config)) || {}
      const defaultValue = resolved.defaultValue || getDefaultValueFromPrototype(prototype, key)
      const type = resolveType(resolved.type, prototype, key)
      return {...resolved, defaultValue, type, name: key}
    })
  }
}

// Needs to be exported so Arg can use it
export interface InputFieldDefinitions {
  <TValue extends string | number | boolean>(config?: Thunk<PrimitiveInputFieldConfig<TValue>>): <
    TName extends string
  >(
    prototype: Record<TName, string> | Record<TName, boolean> | Record<TName, number>,
    key: TName
  ) => void

  <TValue>(config: Thunk<InputFieldOptions<TValue>>): <TName extends string>(
    prototype: Record<TName, TValue>,
    key: TName
  ) => void
}

export const InputField = InputFieldDecorator as InputFieldDefinitions
