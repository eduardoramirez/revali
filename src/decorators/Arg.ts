import 'reflect-metadata';
import InputField from './InputField';

export default InputField;
