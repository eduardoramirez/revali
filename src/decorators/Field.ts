import {GraphQLInputType, GraphQLOutputType, GraphQLResolveInfo, Thunk} from 'graphql'

import {InterfaceImplementation} from 'revali/decorators/Implements'
import {graph} from 'revali/graph'
import {AnyConstructor, EmptyConstructor, MaybePromise} from 'revali/types'
import {resolveThunk, resolveType} from 'revali/utils'
import {WrapperOrType} from 'revali/wrappers/Wrapper'

export interface FieldOptions<TReturn, TArgs = {}, TContext = any> {
  type: WrapperOrType<TReturn, GraphQLOutputType>
  description?: string
  args?: EmptyConstructor<TArgs>
  arg?: WrapperOrType<TArgs, GraphQLInputType>
  deprecationReason?: string
  context?: AnyConstructor<TContext>
}

export type FieldResolverMethod<TContext, TReturn, TArgs> = (
  args: TArgs,
  context: TContext,
  info: GraphQLResolveInfo
) => MaybePromise<TReturn | InterfaceImplementation<TReturn>>

export type FieldProperty<TContext, TReturn, TArgs> =
  | MaybePromise<TReturn>
  | FieldResolverMethod<TContext, TReturn, TArgs>

function FieldDecorator<TReturn, TArgs, TContext>(
  config?: Thunk<Partial<FieldOptions<TReturn, TArgs, TContext>>>
) {
  return <TName extends string, TRArgs extends TArgs = TArgs>(
    prototype: Record<TName, FieldProperty<TContext, TReturn, TRArgs>>,
    key: TName
  ) => {
    graph.createField(prototype.constructor, () => {
      const resolved = (config && resolveThunk(config)) || {}

      const type = resolveType(resolved.type, prototype, key)

      return {...resolved, type, name: key}
    })
  }
}

export function fieldDecoratorForContext<TContext>(context: AnyConstructor<TContext>) {
  return <TReturn, TArgs>(config?: Thunk<Partial<FieldOptions<TReturn, TArgs>>>) => {
    return FieldDecorator<TReturn, TArgs, TContext>({...config, context})
  }
}

type FieldPropertyDecorator<TReturn, TArgs, TContext = undefined> = <
  TName extends string,
  TSource,
  TRArgs extends TArgs = TArgs
>(
  prototype: Record<TName, FieldProperty<TContext, TReturn, TArgs>>,
  key: TName
) => void

interface FieldDefinitions {
  // Infer primitive types. Will not work on methods since the type for a method is,
  // for instance, `() => string`
  <TArgs>(config?: Thunk<Partial<FieldOptions<undefined, TArgs>>>): <
    TName extends string,
    TSource,
    TContext,
    TRArgs extends TArgs = TArgs
  >(
    prototype: Record<TName, string> | Record<TName, boolean> | Record<TName, number>,
    key: TName
  ) => void

  <TReturn, TArgs, TContext = undefined>(
    config: Thunk<FieldOptions<TReturn, TArgs, TContext>>
  ): FieldPropertyDecorator<TReturn, TArgs, TContext>
}

export const Field = FieldDecorator as FieldDefinitions
