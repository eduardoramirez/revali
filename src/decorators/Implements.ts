import { AbstractConstructor, Constructor } from '../types';
import { FieldProperty } from './Field';

type InterfaceImplementation<T> = {
  [key in keyof T]: FieldProperty<any, T[key], any>
}

export default <TIface>(iface: AbstractConstructor<TIface>) =>
  <TType extends InterfaceImplementation<TIface>>(target: Constructor<TType>) => {

  }
