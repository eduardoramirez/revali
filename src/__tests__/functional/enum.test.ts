import {getIntrospectionQuery, graphql, printSchema} from 'graphql'

import {ObjectType} from 'revali/decorators'
import {Arg, compileSchema, enumType, Field, InputField, InputObjectType} from 'revali/index'
import {TSGraphQLBoolean} from 'revali/wrappers/scalars'

describe('Enums', () => {
  it('generates correct output for enums', async () => {
    enum NumberEnum {
      One = 1,
      Two,
      Three,
      Four,
    }
    const NumberEnumType = enumType(NumberEnum, {name: 'NumberEnum'})

    enum StringEnum {
      One = 'ONE',
      Two = 'TWO',
      Three = 'THREE',
    }
    const StringEnumType = enumType(StringEnum, {
      name: 'StringEnum',
      description: 'custom string enum',
    })

    @InputObjectType()
    class NumberEnumInput {
      @InputField({type: NumberEnumType})
      public numberEnumField!: NumberEnum
    }

    @InputObjectType()
    class StringEnumInput {
      @InputField({type: StringEnumType})
      public stringEnumField!: StringEnum
    }

    @ObjectType()
    class Query {
      @Field({type: NumberEnumType, arg: NumberEnumInput})
      public getNumberEnumValue(@Arg('input') input: NumberEnumInput): NumberEnum {
        return NumberEnum.Two
      }

      @Field({type: StringEnumType, arg: StringEnumInput})
      public getStringEnumValue(@Arg('input') input: StringEnumInput): StringEnum {
        return StringEnum.Two
      }

      @Field({type: TSGraphQLBoolean, arg: NumberEnumType})
      public isNumberEnumEqualOne(@Arg('enum') numberEnum: NumberEnum): boolean {
        return numberEnum === NumberEnum.One
      }

      @Field({type: TSGraphQLBoolean, arg: StringEnumType})
      public isStringEnumEqualOne(@Arg('enum') stringEnum: StringEnum): boolean {
        return stringEnum === StringEnum.One
      }
    }

    const schema = compileSchema({Query})
    const {errors} = await graphql(schema, getIntrospectionQuery())
    expect(errors).toBeFalsy()
    expect(printSchema(schema)).toMatchSnapshot()
  })
})
