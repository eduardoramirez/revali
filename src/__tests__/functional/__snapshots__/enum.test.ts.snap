// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Enums generates correct output for enums 1`] = `
"enum NumberEnum {
  One
  Two
  Three
  Four
}

input NumberEnumInput {
  numberEnumField: NumberEnum!
}

type Query {
  getNumberEnumValue(input: NumberEnumInput!): NumberEnum!
  getStringEnumValue(input: StringEnumInput!): StringEnum!
  isNumberEnumEqualOne(enum: NumberEnum!): Boolean!
  isStringEnumEqualOne(enum: StringEnum!): Boolean!
}

\\"\\"\\"custom string enum\\"\\"\\"
enum StringEnum {
  One
  Two
  Three
}

input StringEnumInput {
  stringEnumField: StringEnum!
}
"
`;
